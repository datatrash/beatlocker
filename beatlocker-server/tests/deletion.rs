use crate::test_utils::{copy_recursively, TestClient};
use axum::http::StatusCode;
use beatlocker_server::*;
use chrono::{DateTime, Utc};
use std::path::PathBuf;
use std::sync::Arc;

#[path = "test_utils/mod.rs"]
mod test_utils;

use test_utils::*;

#[tokio::test]
async fn deletion_test() -> AppResult<()> {
    let tempdir = tempfile::Builder::new().rand_bytes(0).tempdir_in(".")?;
    copy_recursively("tests/data", tempdir.path())?;

    enable_default_tracing();
    let options = ServerOptions {
        path: tempdir.path().to_path_buf(),
        database: DatabaseOptions {
            path: Some(PathBuf::from(".")),
            in_memory: true,
        },
        now_provider: Arc::new(Box::new(|| {
            DateTime::parse_from_rfc3339("2020-02-02T00:00:00Z")
                .unwrap()
                .with_timezone(&Utc)
        })),
        ..Default::default()
    };
    let app = App::new(options).await?;
    let client = TestClient::new(app.app.clone());

    app.task_manager
        .send(app.import_all_folders().await?)
        .await?;

    std::fs::remove_file(
        tempdir
            .path()
            .join("Richard Bona/Richard Bona - Ba Senge.ogg"),
    )?;

    app.task_manager
        .send(app.remove_deleted_files().await?)
        .await?;

    let res = client
        .get(&format!("/rest/getArtist?f=json&id={RICHARD_BONA_UUID}"))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getArtistAfterDeletingBaSenge.json",
        res.json::<serde_json::Value>().await
    );

    std::fs::remove_dir_all(tempdir.path().join("Richard Bona"))?;
    app.task_manager
        .send(app.remove_deleted_files().await?)
        .await?;

    let res = client
        .get(&format!("/rest/getAlbumList?f=json"))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getAlbumListAfterDeletingAlbum.json",
        res.json::<serde_json::Value>().await
    );
    let res = client.get(&format!("/rest/getArtists?f=json")).send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getArtistsAfterDeletingAlbum.json",
        res.json::<serde_json::Value>().await
    );

    Ok(())
}
