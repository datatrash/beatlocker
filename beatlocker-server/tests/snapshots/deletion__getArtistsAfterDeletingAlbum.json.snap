---
source: beatlocker-server/tests/deletion.rs
assertion_line: 75
expression: "res.json::<serde_json::Value>().await"
---
{
  "subsonic-response": {
    "artists": {
      "index": [
        {
          "artist": [
            {
              "albumCount": 1,
              "id": "c2042f2f-fbda-64e4-ff33-62bad6853d99",
              "name": "Alex Gopher"
            }
          ],
          "name": "A"
        },
        {
          "artist": [
            {
              "albumCount": 0,
              "id": "a597d760-ecda-330c-8e48-b8a92ba19a25",
              "name": "Unknown Artist"
            }
          ],
          "name": "U"
        }
      ]
    },
    "ignoredArticles": "The El La Los Las Le Les Os As O A",
    "serverVersion": "unknown",
    "status": "ok",
    "type": "beatlocker",
    "version": "1.16.1"
  }
}
